devito/types.py

```diff
@@ -622,9 +622,18 @@ def indexify(self, indices=None):
         """Create a :class:`sympy.Indexed` object from the current object."""
         if indices is not None:
             return Indexed(self.indexed, *indices)
-        # Only replace spacing -> 1 if used as index
-        subs = dict([(i.spacing, 1) for i in self.indices if
-                     any(i in a.args for a in self.args)])
+
+        # Get spacing symbols for replacement
+        spacings = [i.spacing for i in self.indices]
+
+        # Only keep the ones used as indices.
+        spacings = [s for i, s in enumerate(spacings)
+                    if tuple(s.free_symbols)[0] in self.args[i].free_symbols]
```

Comment:
why u taking the first entry in ``tuple(s.free_symbols)``?
isn't ``tuple(s.free_symbols)[0] == s`` ?

Reply:
`s.free_symbols` is a set

Reply:
ah, then use intersection